メッセージ送信機能の実装

norika69 opened the pull request:
## what
メッセージを送信できる
メッセージの本文が空の時保存されないようにする
保存できない理由をフラッシュメッセージで出す

## why
chat-spaceにてメッセージ送信を可能にするため

norika69 committed:
edit messages.scss

messages.scssの調整

norika69 committed:
add create action to messages_controller

グループテーブルからparamsを探してくる
@messageにgroupテーブルのidに関連付けしてmessage_paramsのストロングパラメータをかけて入力されたmessageのカラ
ムを作る
バリテーション確認
もし＠messageに保存されたらshowにリデイレクト
違ったらshowに飛んでエラー文を出す
（.full_messages[0]はエラー文を全部表示し、[0]は括弧をなくす）

norika69 committed:
add show action to groups_controller

groupテーブルからカラムを探す
探してきたカラムのidのmessageを@messagesに格納

norika69 committed:
add validates to message model

norika69 committed:
edit show.html.haml&config/app.rb

メッセージ一覧の表示
= message.created_at.strftime("%Y/%m/%d %H:%M:%S")
は正規表現を用いて投稿日時の年月日時刻まで表示。strftimeは日本時刻に変換

norika69 committed:
edit ja.yml

エラー文の編集

norika69 committed:
add users_cotroller.rb

不要なコードの削除

norika69 committed:
edit edit.html.haml

不要なコードの削除

hand12 commented on app/controllers/groups_controller.rb:
```diff
@@ -20,6 +20,8 @@ def create
   def show
     @message = Message.new
+    @group = Group.find(params[:id])
+    @messages = @group.messages
```
こちら、set_groupの中で定義されており、before_actionによりshowが呼ばれる前にset_groupメソッドが呼ばれ、@groupは定義されていますので、こちらの記述はなくて良さそうですね！

hand12 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,5 +1,34 @@
+  end
+
+  def set_message
+    @group = Group.find(params[:id])
```
set_messageメソッドを定義しているのですが、実際にやっていることはgroupを@groupをセットしていますね。
名前を変更した方がいいですね！
また、せっかくこのメソッドを定義したのに呼び出す処理が書いていないので、呼び出す処理も書いてあげましょう！

hand12 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,5 +1,34 @@
+    if @message.save
+       redirect_to group_path(params[:id])
+    else
+       redirect_to group_path(params[:id]),alert: @message.errors.full_messages[0]
```
redirect_toとrenderの違いを調べてみてください！

norika69 replied:
redirect_toはコントローラーのアクションを通してビューファイルを表示します。
それに対し、renderは直接ビューファイルを返すので、rengerの方が早く通信ができます。
使い分けとしては、redirect_toはデータの追加、更新、削除するときに、rendarはただデータを取得して表示したいときに使います。

hand12 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,5 +1,34 @@
+
+  def destroy
+
+  end
```
このプルリクでは使っていないアクションは今定義しなくて良さそうですね！
削除しておきましょう！

hand12 commented on app/controllers/users_controller.rb:
```diff
@@ -1,4 +1,2 @@
 class UsersController < ApplicationController
-  belongs_to :user
-  belongs_to :group
 end
```
usersコントローラーも使っていないのであれば削除しましょう！

hand12 commented on app/views/groups/edit.html.haml:
```diff
@@ -1,5 +1,4 @@
-  = @users
   .chat-group-form
     %h1  チャットグループ編集
     = form_for(@group, html: {class: "edit_chat_group", id: ""}) do |f|
```
idなど指定していないのであれば、書かなくて大丈夫です。

hand12 commented on app/views/groups/show.html.haml:
```diff
@@ -6,31 +6,34 @@
         = link_to '', new_group_path, class: "fa fa-edit icon-edit"
         = link_to '', edit_user_registration_path(current_user), class: "fa fa-cog icon-cog"
-    = render 'chat_left_group', object: current_user.groups
+    = render 'chat_left_group',object: current_user.groups
```
objectは、ファイル名と同じ名前の変数を用意し、指定した値をその変数の中にいれて部分テンプレートの中でも使うというものです。
これは他にもやり方があるかつ、変数名がわかりづらくなるので、部分テンプレートに変数を渡す違うやり方を調べてみてください！

hand12 commented on app/views/groups/show.html.haml:
```diff
@@ -6,31 +6,34 @@
+        = link_to 'Edit', edit_group_path(), class: "editbutton"
       %p
-        Member:
+        = @group.users
```
ここエラー起きてないですか？

norika69 replied:
修正しました。

hand12 commented on app/views/groups/show.html.haml:
```diff
@@ -6,31 +6,34 @@
-              2016/11/18
-          %p.chat__main__body__messages__message__body
-            text
+        -@messages.each do |message|
```
-の後も半角スペース空けましょう！

hand12 replied:
うまく動いているのが確認できたら、今度はこちらを部分テンプレートにしてみましょう！
部分テンプレートの方が繰り返し処理は早いです。

hand12 reviewed (commented):
いくつかコメントしたのでご確認ください！

norika69 committed:
edit groups_controller

before_actionで定義しているため、削除

norika69 committed:
edit messages_controller

before_actionの設定
不要なアクションの削除
　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　
　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　
　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　
　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　　

norika69 committed:
delete users_controller

norika69 committed:
create messages parcial file

パーシャルファイルを作成して、メッセージ一覧部分を一括除去

norika69 committed:
edit messages.scss

ビューファイルの調整

norika69 committed:
edit show.html.haml&_chat_left_group.html.haml

render :objectオプションから
:partial,localesオプションへの変更

norika69 committed:
edit _messages.html.haml

インデントの修正

norika69 committed:
edit messages_controller

redirect_to→render

hc0208 commented on app/controllers/messages_controller.rb:
```diff
@@ -1,5 +1,38 @@
+  end
+
+  def set_group
+    @group = Group.find(params[:id])
```
全体的にインデントがおかしいので整えてあげるようにしましょう。

hc0208 commented on app/views/groups/show.html.haml:
```diff
@@ -6,31 +6,31 @@
-          %form{ action: "#", method: "post"}
-            %textarea{ placeholder: "type a massage" , rows: "1", id: "message-body"}
-            %label.chat__main__footer__form__body__label
+          = form_for @message, :url => messages_path do |f|
```
url: messages_pathと書くことが多いです。

hc0208 commented on app/views/groups/show.html.haml:
```diff
@@ -6,31 +6,31 @@
-            %label.chat__main__footer__form__body__label
+          = form_for @message, :url => messages_path do |f|
+            = f.text_area :body, placeholder: 'type a massage',rows: 1, id: "message-body"
+            = f.hidden_field :user_id, :value => current_user.id
```
value: current_user.idと書くことが多いです。

norika69 committed:
edit messages_controller

インデントの修正

norika69 committed:
edit show.html.haml

ハッシュロケットからコロンへ修正

hc0208 commented:
LGTM
